VMware: Fix module vmware_host

ckotte opened the pull request:
##### SUMMARY
- An ESXi host cannot be added, removed, or re-connected if it's located under a folder instead of a cluster
- Fixes https://github.com/ansible/ansible/issues/44772 if you connect the ESXi host under a datacenter
- moves the host to a new location (from cluster to folder, or vice versa)
- fetches SSL thumbprint automatically, if `ssl_thumbprint` isn't set, without executing the tasks again if failed because of SSL verify fault
- connects the host if it's disconnected (enabled by default)

##### ISSUE TYPE
- Bugfix Pull Request
- Feature Pull Request

##### COMPONENT NAME
vmware_host

##### ANSIBLE VERSION
```
# ansible --version
ansible 2.6.3
  config file = /root/ansible-vmware/ansible.cfg
  configured module search path = [u'/root/ansible-vmware/library']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.5 (default, Jul 13 2018, 13:06:57) [GCC 4.8.5 20150623 (Red Hat 4.8.5-28)]
```

##### ADDITIONAL INFORMATION
before:
```
TASK [esxi : Add ESXi host to vCenter] *********************************************************************************************************************************************************************************************
fatal: [esxi_host -> jump_server]: FAILED! => {"changed": false, "msg": "Failed to add host esxi_host to vCenter: (vim.fault.DuplicateName) {\n   dynamicType = <unset>,\n   dynamicProperty = (vmodl.DynamicProperty) [],\n   msg = \"The name 'esxi_host' already exists.\",\n   faultCause = <unset>,\n   faultMessage = (vmodl.LocalizableMessage) [],\n   name = 'esxi_host',\n   object = 'vim.HostSystem:host-2640'\n}"}
```
```
TASK [esxi : Remove ESXi host from vCenter] *********************************************************************************************************************************************************************************************
fatal: [esxi_host -> jump_server]: FAILED! => {"changed": false, "msg": "'vim.ComputeResource' object has no attribute 'runtime'"}
```
```
TASK [esxi : Reconnect ESXi host] *********************************************************************************************************************************************************************************************
fatal: [esxi_host -> jump_server]: FAILED! => {"changed": false, "msg": "Failed to reconnect host esxi_host to vCenter: A specified parameter was not correct: cnxSpec"}
```
after:
```paste below
TASK [esxi : Add ESXi host to vCenter] *******************************************************************************************************************************************************************************************************
changed: [esxi_host -> jump_server]
```

ckotte commented:
> The test `ansible-test sanity --test validate-modules` [[explain](https://docs.ansible.com/ansible/devel/dev_guide/testing/sanity/validate-modules.html)] failed with 1 error:
> 
> ```
> lib/ansible/modules/cloud/vmware/vmware_host.py:0:0: E309 version_added for new option (folder_name) should be 2.8. Currently 2.6
> ```

Changed version_added from 2.6 to 2.8. Looks like adding an ESXi host to a folder didn't even work in 2.6.

Akasurde commented:
> > The test `ansible-test sanity --test validate-modules` [[explain](https://docs.ansible.com/ansible/devel/dev_guide/testing/sanity/validate-modules.html)] failed with 1 error:
> > ```
> > lib/ansible/modules/cloud/vmware/vmware_host.py:0:0: E309 version_added for new option (folder_name) should be 2.8. Currently 2.6
> > ```
> 
> Changed version_added from 2.6 to 2.8. Looks like adding an ESXi host to a folder didn't even work in 2.6.

No, error is suggesting that you are adding new parameter to Ansible module, currently documentation is indicating that it is being added in 2.6 but current development cycle is 2.8. 

I would recommend to add new `aliases` rather than renaming parameter altogether for backward compatibility.

ckotte commented:
> I would recommend to add new `aliases` rather than renaming parameter altogether for backward compatibility.

Ok. Makes sense. I will do that.

I still think that the folder feature didn't work. There was also no test created for it. I will add that too once I figured out how to do it...

ckotte committed:
Add folder_name alias

The parameter should be named folder_name instead of folder to be
consistent with the other parameters cluster_name and datacenter_name.

ckotte committed:
Fix 'A specified parameter was not correct: cnxSpec' error

The play fails when re-connecting a host with the cnxSpec argument
defined.

Just remove the cnxSpec argument definition. It isn't necessary when
re-connecting an existing host in vCenter (6.5).

ckotte committed:
Fix ESXi host add/remove/re-connect errors

An ESXi host cannot be added, removed, or re-connected under a folder.

ESXi hosts connected under a folder, instead of a cluster, have an
object type of ComputeResource instead of HostSystem and the HostSystem
object is the first entry in the host parameter list of the
ComputeResource object.

Separate between cluster and folder actions.

mattclay commented:
CI failure in integration tests: https://app.shippable.com/github/ansible/ansible/runs/86604/86/tests

ckotte commented:
> CI failure in integration tests: https://app.shippable.com/github/ansible/ansible/runs/86604/86/tests

I know. Looks like you cannot add an ESXi host to a folder with vcsim. I need more time to investigate.

ckotte commented:
It's not possible to test the connection of an ESXi host to a folder with vcsim.

tchernomax commented on lib/ansible/modules/cloud/vmware/vmware_host.py:
```diff
@@ -50,11 +50,12 @@
     - "  - '/Site2/dc2/Asia-Cluster/host'"
     - "  - '/dc3/Asia-Cluster/host'"
     version_added: "2.6"
+    aliases: ['folder_name']
```
I am ok to change "folder" to "folder_name" as the option for the datacenter name is "datacenter_name".
But if you do, do it everywhere (option name and python variable) and add an alias "folder" for retro-compatibility.

dagwieers replied:
I would leave it like this. I don't see what `_name` adds except maybe confusion. Obviously it's the folder you have to provide. I would leave it at `C(folder)` in the description.

tchernomax replied:
that's also ok for me to leave it like this

ckotte replied:
The "problem" with all the vmware modules is that they are not consistent. One module uses e.g. datacenter_name and the other uses only datacenter and so on.

dagwieers replied:
I think we need to have consistency for all the VMware-related modules wrt. this. And my proposal is to not use `_name` for the standard parameters. (We can have them as aliases though)

dagwieers replied:
I will put it on the agenda for next week's meeting.

ckotte replied:
Ok. I had the impression that parameters with _name are more used (datacenter _name, cluster_name, switch_name, etc. pp. ). I'm fine if we can achieve consistency.So, for example, you recommend to change datacenter_name to datacenter and (only) add an alias for backward compatibility?

dagwieers replied:
I think that's how it starts to diverge, impressions of what other people use ;-)
No worries, a standard is what we need, and any standard will do. Join the next VMware meeting (ansible/community#233) to voice your opinion !

ckotte replied:
Changed back to folder, but leave alias.

tchernomax commented on lib/ansible/modules/cloud/vmware/vmware_host.py:
```diff
@@ -283,11 +284,13 @@ def reconnect_host_to_vcenter(self):
+                        task = self.host.host[0].ReconnectHost_Task(**reconnecthost_args)
+                    elif self.cluster:
+                        task = self.host.ReconnectHost_Task(**reconnecthost_args)
                     success, result = wait_for_task(task)
```
IMHO it would be simpler to set self.host as a HostSystem object.
And add a self.host_parent_compute_resource variable for the parent ComputeResource (when self.folder).

the host_parent_compute_resource would only be used for Destroy_Task as: https://www.vmware.com/support/developer/vc-sdk/visdk25pubs/ReferenceGuide/vim.HostSystem.html
> Invoking destroy on a HostSystem of standalone type throws a NotSupported fault. A standalone HostSystem can be destroyed only by invoking destroy on its parent ComputeResource.

tchernomax replied:
@ckotte What do you think ?

ckotte replied:
> @ckotte What do you think ?

Yes, I haven't had time for this yet.  Could you provide me some example code?

tchernomax replied:
@ckotte https://paste.gnome.org/pck6vwny2

ckotte replied:
Thx. I needed to update the remove method as well.
```
if self.folder:
    task = self.host_parent_compute_resource.Destroy_Task()
elif self.cluster:
    task = self.host.Destroy_Task()
changed, result = wait_for_task(task)
```
(You mentioned this in the first post already)

ckotte commented:
Ok. I had the impression that parameters with _name are more used (datacenter _name, cluster_name, switch_name, etc. pp. ). I'm fine if we can achieve consistency.So, for example, you recommend to change datacenter_name to datacenter and (only) add an alias for backward compatibility?

tchernomax commented:
@ckotte maybe you can separate this PR in two:
* rename
* fix

ckotte commented:
> @ckotte maybe you can separate this PR in two:
> 
> * rename
> * fix

I just added an alias on top of the fix. I don't want to create and manage another branch and PR just for this little fix. I also use that module already in my "production" playbook and I want to use `folder_name` because almost everything other parameter also uses `_name`.

"They" just need to establish a "vmware-related module standard parameter (and alias) naming scheme".

tchernomax reviewed (approved):
LGTM

ckotte commented:
@tchernomax @Akasurde Could you please review the module again? I added update functionality and I needed to basically refactor the module.

corbinstuard commented:
What's holding this PR up? I'm currently experiencing the issue where adding a host that is already connected errors out instead or returning 'ok':

```
The full traceback is:
WARNING: The below traceback may *not* be related to the actual failure.
  File "/tmp/ansible_vmware_host_payload_vglXIB/__main__.py", line 258, in add_host_to_vcenter
    success, result = wait_for_task(task)
  File "/tmp/ansible_vmware_host_payload_vglXIB/ansible_vmware_host_payload.zip/ansible/module_utils/vmware.py", line 68, in wait_for_task
    raise_from(TaskError(error_msg, host_thumbprint), task.info.error)
  File "/tmp/ansible_vmware_host_payload_vglXIB/ansible_vmware_host_payload.zip/ansible/module_utils/six/__init__.py", line 744, in raise_from
    raise value

fatal: [10.1.2.71 -> localhost]: FAILED! => {
    "changed": false,
    "invocation": {
        "module_args": {
            "add_connected": true,
            "cluster_name": null,
            "datacenter_name": "dispersive",
            "esxi_hostname": "10.1.2.71",
            "esxi_password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "esxi_ssl_thumbprint": "",
            "esxi_username": "root",
            "folder": "/dispersive/host",
            "hostname": "10.1.2.19",
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "port": 443,
            "state": "present",
            "username": "administrator@vsphere.local",
            "validate_certs": false
        }
    },
    "msg": "Failed to add host 10.1.2.71 to vCenter: The name '10.1.2.71' already exists."
```

ckotte commented:
@akashihi Could you please approve this PR? I don't get any review since October 2018!!

Akasurde commented:
Closing and re-opening for CI trigger.

Akasurde commented:
@ckotte Thanks for the contribution.
